# 💡 예외 계층
* 스프링이 제공하는 예외 추상화를 이해하기 위해서는 먼저 자바 기본 예외에 대한 이해가 필요하다.

![image](https://github.com/shin-je-woo/TIL/assets/39439576/b47698cb-ffb8-48ab-aa9b-4f175721ac5e)
* `Object` : 예외도 객체이다. 모든 객체의 최상위 부모는 Object 이므로 예외의 최상위 부모도 Object이다.
* `Throwable` : 최상위 예외이다. 하위에 Exception 과 Error 가 있다.
* `Error` : 메모리 부족이나 심각한 시스템 오류와 같이 애플리케이션에서 복구 불가능한 시스템 예외이다. 
  * 애플리케이션 개발자는 이 예외를 잡으려고 해서는 안된다.
  * 상위 예외를 catch 로 잡으면 그 하위 예외까지 잡기 때문에 Throwable 예외도 잡으면 안된다. (Error 예외도 함께 잡을 수 있기 때문이다.)
  * 애플리케이션 로직은 이런 이유로 Exception 부터 필요한 예외로 생각하고 잡으면 된다. 참고로 Error 도 언체크 예외이다.
* `Exception` : 체크 예외
  * 애플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외이다.
  * Exception 과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외이다. 단 RuntimeException 은 예외로 한다.
* `RuntimeException` : 언체크 예외, 런타임 예외
  * 컴파일러가 체크 하지 않는 언체크 예외이다.
  * RuntimeException 과 그 자식 예외는 모두 언체크 예외이다.
  * RuntimeException 의 이름을 따라서 RuntimeException 과 그 하위 언체크 예외를 런타임 예외라고 많이 부른다.

## 예외 기본 규칙
* 예외는 폭탄 돌리기와 같다. 잡아서 처리하거나, 처리할 수 없으면 밖으로 던져야한다.
* 예외에 대해서는 2가지 기본 규칙을 기억하자.
1. 예외는 잡아서 처리하거나 던져야 한다.
2. 예외를 잡거나 던질 때 지정한 예외뿐만 아니라 그 예외의 자식들도 함께 처리된다.
    * 예를 들어서 `Exception` 을 `catch` 로 잡으면 그 하위 예외들도 모두 잡을 수 있다.
    * 예를 들어서 `Exception` 을 `throws` 로 던지면 그 하위 예외들도 모두 던질 수 있다

#### 참고)  예외를 처리하지 못하고 계속 던지면 어떻게 될까?
* 자바 main() 쓰레드의 경우 예외 로그를 출력하면서 시스템이 종료된다.
* 웹 애플리케이션의 경우 여러 사용자의 요청을 처리하기 때문에 하나의 예외 때문에 시스템이 종료되면 안된다. 
* WAS가 해당 예외를 받아서 처리하는데, 주로 사용자에게 개발자가 지정한 오류 페이지를 보여준다.

# 💡 체크 예외 기본 이해
* `Exception` 과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외이다. 단, `RuntimeException` 은 예외로 한다.
* 체크 예외는 잡아서 처리하거나, 또는 밖으로 던지도록 선언해야한다. 그렇지 않으면 컴파일 오류가 발생한다.
* `Exception` 을 상속 받으면 체크 예외가 된다. 참고로 `RuntimeException` 을 상속 받으면 언체크 예외가 된다.
```java
...
public void callCatch() {
    try {
        repository.call();
    } catch (MyCheckedException e) {
        //예외 처리 로직
        log.info("예외 처리, message={}", e.getMessage(), e);
    }
}
...
public void callThrow() throws MyCheckedException {
    repository.call();
}
...
static class Repository {
    public void call() throws MyCheckedException {
        throw new MyCheckedException("ex");
    }
}
```
* 체크 예외를 잡아서 처리하려면 `catch(..)` 를 사용해서 예외를 잡으면 된다.
* 여기서는 `MyCheckedException` 예외를 잡아서 처리한다.
* log를 남길 때 로그의 마지막 인수에 예외 객체를 전달해주면 로그가 해당 예외의 스택 트레이스를 추가로 출력해 준다.
* catch는 해당 타입과 그 하위 타입을 모두 잡을 수 있다.
* 따라서 `MyCheckedException` 을 상위 타입인 `Exception` 으로 바꿔도 예외를 잡을 수 있다.
* 체크 예외를 처리할 수 없을 때는 `method() throws 예외` 를 사용해서 밖으로 던질 예외를 필수로 지정해주어야 한다. (체크 예외를 잡지도, 던지지도 않으면 컴파일 오류가 발생한다.)
* 여기서는 callThrow() 메서드에서 `MyCheckedException` 을 밖으로 던지도록 지정해주었다.
* `throws` 에 `MyCheckedException` 의 상위 타입인 `Exception` 을 적어 주어도 `MyCheckedException` 을 던질 수 있다.
* **장점**
  * 개발자가 실수로 예외를 누락하지 않도록 컴파일러를 통해 문제를 잡아주는 훌륭한 안전 장치이다.
* **단점**
  * 개발자가 모든 체크 예외를 반드시 잡거나 던지도록 처리해야 하기 때문에, 너무 번거로운 일이 된다. 
  * 크게 신경쓰고 싶지 않은 예외까지 모두 챙겨야 한다. 추가로 의존관계에 따른 단점도 있다.

# 💡 언체크 예외 기본 이해
* `RuntimeException` 과 그 하위 예외는 언체크 예외로 분류된다.
* 언체크 예외는 말 그대로 컴파일러가 예외를 체크하지 않는다는 뜻이다.
* 언체크 예외는 체크 예외와 기본적으로 동일하다. 차이가 있다면 예외를 던지는 `throws` 를 선언하지 않고 생략할 수 있다. 이 경우 자동으로 예외를 던진다.

#### 체크 예외 VS 언체크 예외
* 체크 예외: 예외를 잡아서 처리하지 않으면 항상 `throws` 에 던지는 예외를 선언해야 한다.
* 언체크 예외: 예외를 잡아서 처리하지 않아도 `throws` 를 생략할 수 있다.
